src/azul/openapi/validation.py

```diff
@@ -0,0 +1,218 @@
+import json
+import logging
+from typing import (
+    Callable,
+    Mapping,
+    Optional,
+)
+
+import jsonschema
+from more_itertools import (
+    one,
+)
+
+from azul import (
+    RequirementError,
+    reject,
+)
+from azul.openapi.schema import (
+    type_lookup,
+)
+from azul.plugins.metadata.hca.transform import (
+    value_and_unit,
+)
+from azul.types import (
+    JSON,
+)
+from azul.uuids import (
+    InvalidUUIDError,
+    InvalidUUIDVersionError,
+    validate_uuid,
+)
+
+logger = logging.getLogger(__name__)
+
+
+class ValidationError(Exception):
+
+    def __init__(self,
+                 message,
+                 name,
+                 value,
+                 schema: Optional[JSON] = None):
+        self.message = message
+        self.name = name
+        self.value = value
+        if schema is not None:
+            self.schema = schema
+
+
+class SpecValidator:
+
+    def __init__(self,
+                 *,
+                 name: str,
+                 required: bool,
+                 default: str = None,
+                 schema: Mapping[str, str] = None,
+                 content: Mapping[str, str] = None,
+                 pattern: Optional[str] = None,
+                 validator: Callable,
+                 **kwargs):
+        """
+        :param kwargs: Since 'in' is a Python reserved word we can only extract it
+         using kwargs. All other kwargs are ignored.
+        """
+        self.in_ = kwargs['in']
+        self.json_validators = dict(jsonschema.Draft4Validator.VALIDATORS)
+        self.name = name
```

Comment:
Replace this boiler-plate with `attr.s`

Reply:
Jesse and I spent some time looking at this, the issue we found was that the parameter `in` is used within the schema that defines the parameter: https://swagger.io/specification/#parameterObject . I'll take another look, not sure if it can be done. https://github.com/python-attrs/attrs/issues/381